src/datepicker/utils/types.js

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright (c) 2018-2020 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+
+type DateValues<T> = T | string | number;
+type DateInput<T> = ?DateValues<T>;
+
+type Comparison<T> = (value: T, comparing: T) => boolean;
+type DateInDateOut<T> = (value: T) => T;
+
+type DateFuncT<T> = (DateInput<T> | void) => T;
```

Comment:
It can, but only if we you pass null, and since we're using the types to prevent people from passing null, this seems like a reasonable compromise.  If we don't do this, every single time we call `adapter.date()` we have to do `adapter.date() || new Date` even though the second part of that expression can never happen.  This was that whole predicate function issue I was talking about before.

Reply:
Does this line error if it does not fallback to `undefined`?
https://github.com/uber/baseweb/pull/2968/files#diff-a551f5d9fe815b88dbfdfccf4f31a6afR168

Reply:
This required making the types a little bit more specific, but now it does show a helpful message